### DISCUSSION TITLE
Save time-evolution results of PDE

### DISCUSSION THREAD

**Discussion by pmistriotis** [zwicker-group/py-pde#151] [Q&A] [answered]
Hello, 

First of all, kudos to the pypde team for this amazing package!

I am solving the Brusselator model at pypde and I would like to track and save the entire time-evolution of u,v. Perhaps save it in the 'sol._data' instead of just saving the final snapshot. What should I change in the tracker object to achieve that? 

############
from pde import PDE, FieldCollection, PlotTracker, ScalarField, UnitGrid

### define the PDE
a, b = 1, 3
d0, d1 = 1, 0.1
eq = PDE(
    {
        "u": f"{d0} * laplace(u) + {a} - ({b} + 1) * u + u**2 * v",
        "v": f"{d1} * laplace(v) + {b} * u - u**2 * v",
    }
)

### initialize state
grid = UnitGrid([64, 64])
u = ScalarField(grid, a, label="Field $u$")
v = b / a + 0.1 * ScalarField.random_normal(grid, label="Field $v$")
state = FieldCollection([u, v])

### simulate the pde
tracker = PlotTracker(interval=1, plot_args={"vmin": 0, "vmax": 5})
sol = eq.solve(state, t_range=20, dt=1e-3, tracker=tracker)
#############

**Comment by david-zwicker** [comment, marked as answer]
You can save intermediate data at regular intervals using the [`storage`](https://py-pde.readthedocs.io/en/latest/packages/pde.storage.html) module; see [this example](https://py-pde.readthedocs.io/en/latest/examples_gallery/stochastic_simulation.html) or [this example](https://py-pde.readthedocs.io/en/latest/examples_gallery/pde_1d_expression.html).

**Reply by pmistriotis** [reply]
Thank you for your response!

**Comment by GabrieleNesta** [comment]
Hi, 

sorry for bother you, but I have a question related to this discussion: I am using your extremely useful package to solve a pde by means of a custom class and an explicit solver with adaptive time steps. Is there a way to store exactly what the solver does each time step? 
Since with an adaptive dt I think I cannot know a priori how many time steps there will be and set in advance a storage memory to save them all.

 Lots of thanks for all your help!

**Reply by david-zwicker** [reply]
I'm not sure what your use case is since the post-analysis typically should not depend on the solution details. However, you can implement a `post_step_hook` function to get access to the current state after every time step. It's a bit more of a detail, so it's less well documented, but you should be able to find relevant information by searching the documentation.

**Reply by GabrieleNesta** [reply]
ok, thanks a lot for your answer!!
